File: dg-commons/src/main/code/SocialNetwork_Example/java/UserTransitions.java
package SocialNetwork_Example.java;

import Graph.Node;
import GraphEngine.DataTransitions;
import scala.NotImplementedError;

/**
 * Created by Samer Adra on 7/14/2015.
 */
public class UserTransitions extends DataTransitions<User, UserTypeVal, UserStub, UserTypes> {
    private static UserTransitions ourInstance = new UserTransitions();

    public static UserTransitions getInstance() {
        return ourInstance;
    }

    private UserTransitions() {
    }

    @Override
    public UserTypes nodeDataTypes() {
        return UserTypes.getInstance();
    }

    @Override
    public void linkExistingNodes(Node<User> parent, Node<User> child) {
        parent.addLinkToExistingChild(child);
    }

    @Override
    public boolean isStubParentTypeTransitionAllowed(Node<UserStub> childNode, Node<UserStub> candidateParentNode) {
        UserTypeVal childType = childNode.data().dataType();
        UserTypeVal parentType = candidateParentNode.data().dataType();

        if (childType == UserType.Admin) {
            return parentType == UserType.Admin;
        } else if (childType == UserType.SocialNetworkEmployee) {
            return parentType != UserType.PublicUser;
        } else return true;
    }

    @Override
    public boolean isParentTypeTransitionAllowed(Node<User> childDataNode, Node<User> candidateParentDataNode) {
        throw new NotImplementedError();
    }

    @Override
    public Node<User> addRandomlyGeneratedParentData(Node<User> childDataNode, UserTypeVal parentNodeDataType) {
        throw new NotImplementedError();
    }

    @Override
    public Node<User> addRandomlyGeneratedParentData(Node<User> childDataNode, UserStub parentNodeDataStub) {
        throw new NotImplementedError();
    }

    @Override
    public Node<User> addRandomlyGeneratedChildData(Node<User> parentDataNode, UserTypeVal childNodeDataType) {
        throw new NotImplementedError();
    }

    @Override
    public Node<User> addRandomlyGeneratedChildData(Node<User> parentDataNode, UserStub childNodeDataStub) {
        throw new NotImplementedError();
    }
}

File: dg-commons/src/main/code/SocialNetwork_Example/scala/UserTransitions.scala
package SocialNetwork_Example.scala

import Graph.Node
import GraphEngine.DataTransitions

class UserTransitions

object UserTransitions extends DataTransitions[User, UserType.UserType, UserStub, UserTypes] {
  override def nodeDataTypes = new UserTypes()

  override def isStubParentTypeTransitionAllowed(childNode: Node[UserStub], candidateParentNode: Node[UserStub]): Boolean = {
    childNode.data.dataType match {
      case UserType.Admin =>
        candidateParentNode.data.dataType match {
          case UserType.Admin => true
          case _ => false
        }
      case UserType.SocialNetworkEmployee =>
        candidateParentNode.data.dataType match {
          case UserType.PublicUser => false
          case _ => true
        }
      case UserType.PublicUser => true
    }
  }

  // The following methods all deal with the User object. As we do not yet have a generator that deals with User objects,
  // we leave this to be implemented later. (Currently we have the SocialNetworkStructureBuilder, which builds graphs of
  // UserStubs. So for now we only need the transition methods that take in stubs.)
  override def isParentTypeTransitionAllowed(childDataNode: Node[User], candidateParentDataNode: Node[User]): Boolean = ???
  override def addRandomlyGeneratedParentData(childDataNode: Node[User], parentNodeDataType: UserType.UserType): Node[User] = ???
  override def addRandomlyGeneratedParentData(childDataNode: Node[User], parentNodeDataStub: UserStub): Node[User] = ???
  override def addRandomlyGeneratedChildData(parentDataNode: Node[User], childNodeDataType: UserType.UserType): Node[User] = ???
  override def addRandomlyGeneratedChildData(parentDataNode: Node[User], childNodeDataStub: UserStub): Node[User] = ???
  override def linkExistingNodes(parent: Node[User], child: Node[User]): Unit = {
    parent.addLinkToExistingChild(child)
  }
}
